File: frameworks/rs/tests/java_api/HealingBrush/src/rs/example/android/com/healingbrush/FindRegion.java
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package rs.example.android.com.healingbrush;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.Rect;
import android.graphics.RectF;
import androidx.renderscript.Allocation;
import androidx.renderscript.Element;
import androidx.renderscript.RenderScript;
import androidx.renderscript.Script;
import androidx.renderscript.Type;
import android.util.Log;

import com.example.android.rs.sample.ScriptC_find_region;

public class FindRegion {
    private static final String TAG = "FindRegion";

    Rect mRoiBounds; // bounding box of the ROI
    int[] mPaste; // contains a copy where to paste
    float[] mPointsXY; // polygon point in original image coordnates
    Rect mSearchRange; // range to search in (original image coordinates
    Bitmap mMaskBitmap;

    int mCutOffsetX; // image coords of the cut  (mPointsXY - mPasteOffX + mCutOffsetX)
    int mCutOffsetY; // image coords of the cut (mPointsXY - mPasteOffY + mCutOffsetY)

    public FindRegion(float[] xy, Bitmap img) {
        mPointsXY = xy;

        int imgWidth = img.getWidth();
        int imgHeight = img.getHeight();

        mRoiBounds = getBoundingRect(xy);

        if (mRoiBounds.height() <= 2 ||
                mRoiBounds.left < 0
                || mRoiBounds.top < 0
                || mRoiBounds.right >= imgWidth
                || mRoiBounds.bottom >= imgHeight) {
            throw new RuntimeException("ROI to close to the edge of the image");
        }

        mMaskBitmap = buildMask(mRoiBounds, mPointsXY);
        mPaste = new int[this.mRoiBounds.width() * this.mRoiBounds.height()];
        img.getPixels(mPaste, 0, mRoiBounds.width(),
                mRoiBounds.left, mRoiBounds.top, mRoiBounds.width(), mRoiBounds.height());

        mSearchRange = calcSearchRange(mRoiBounds, imgWidth, imgHeight);
    }

    public int getCutOffsetX() {
        return mCutOffsetX;
    }

    public int getCutOffsetY() {
        return mCutOffsetY;
    }

    public Rect getRoiBounds() {
        return mRoiBounds;
    }


    public Rect findMatch(ScriptC_find_region findRegion, RenderScript mRs, Bitmap image) {
        long time = System.nanoTime();
        Allocation border_coords;
        Allocation border_values;

        Type.Builder builderI32_2 = new Type.Builder(mRs, Element.I32_2(mRs));
        builderI32_2.setX(mPointsXY.length / 2);
        border_coords = Allocation.createTyped(mRs, builderI32_2.create());

        Allocation border_coords_float;
        border_coords_float = Allocation.createSized(mRs, Element.F32_2(mRs), mPointsXY.length / 2);
        border_coords_float.copyFrom(mPointsXY);
        findRegion.forEach_toInt(border_coords_float, border_coords);
        findRegion.set_border_coords(border_coords);

        findRegion.set_image(Allocation.createFromBitmap(mRs, image));

        Type.Builder builderF32_3 = new Type.Builder(mRs, Element.F32_3(mRs));
        builderF32_3.setX(mPointsXY.length / 2);
        border_values = Allocation.createTyped(mRs, builderF32_3.create());
        findRegion.set_border_values(border_values);
        findRegion.forEach_extractBorder(border_coords, border_values);

        Type.Builder builderF32 = new Type.Builder(mRs, Element.F32(mRs));
        builderF32.setX(mSearchRange.width());
        builderF32.setY(mSearchRange.height());

        Allocation fit = Allocation.createTyped(mRs, builderF32.create());
        findRegion.set_borderLength(mPointsXY.length / 2);
        int noSearch_x = mRoiBounds.left - mSearchRange.left;
        int noSearch_y = mRoiBounds.top - mSearchRange.top;
        findRegion.set_imagePosX(noSearch_x);
        findRegion.set_imagePosY(noSearch_y);
        Script.LaunchOptions options = new Script.LaunchOptions();
        options.setX(0, mSearchRange.width() - mRoiBounds.width());
        options.setY(0, mSearchRange.height() - mRoiBounds.height());
        findRegion.forEach_bordercorrelation(fit, options);


        Log.v(TAG, "noSearch " + noSearch_x + ", " + noSearch_y);
        Log.v(TAG, "noSearch " + mRoiBounds.width() + ", " + mRoiBounds.height());

        Allocation fit_max = Allocation.createSized(mRs, Element.I32_2(mRs), 1);

        findRegion.invoke_findMin(fit, fit_max, noSearch_x, noSearch_y,
                mRoiBounds.width(), mRoiBounds.height());
        int[] mina = new int[2];
        fit_max.copyTo(mina);

        mCutOffsetX = mina[0] + mSearchRange.left;
        mCutOffsetY = mina[1] + mSearchRange.top;

        Log.v(TAG, "Time to find replacement= " + (System.nanoTime() - time) / 1E6f + "ms");

        return mRoiBounds;
    }

    /**
     * Computes the bounding box of the polygon
     * then pads and sizes to multiple of 8
     *
     * @param xy points of polygon [x1,y1,x2,y2,...]
     * @return rectangle
     */
    private static Rect getBoundingRect(float[] xy) {
        RectF mRect = calcBounds(xy);
        int mWidth = (((int) (8 + mRect.width())) & ~3); // bounding rectangle that is a power of 8
        int mHeight = (((int) (8 + mRect.height())) & ~3);
        int mPasteOffX = (int) mRect.left - 1;
        int mPasteOffY = (int) mRect.top - 1;
        return new Rect(mPasteOffX, mPasteOffY, mPasteOffX + mWidth, mPasteOffY + mHeight);
    }

    private static RectF calcBounds(float[] xy) {
        float minx = xy[0], miny = xy[1];
        float maxx = xy[0], maxy = xy[1];
        for (int i = 0; i < xy.length; i += 2) {
            minx = Math.min(minx, xy[i]);
            maxx = Math.max(maxx, xy[i]);
            miny = Math.min(miny, xy[i + 1]);
            maxy = Math.max(maxy, xy[i + 1]);
        }
        RectF rect = new RectF();
        rect.set(minx, miny, maxx, maxy);
        return rect;
    }

    private static Bitmap buildMask(Rect rec, float[] xy) {
        Bitmap bitmap = Bitmap.createBitmap(rec.width(), rec.height(), Bitmap.Config.ALPHA_8);

        Canvas c = new Canvas(bitmap);
        Paint paint = new Paint();
        paint.setStyle(Paint.Style.FILL);
        paint.setColor(Color.BLACK);
        Path path = new Path();
        for (int i = 0; i < xy.length; i += 2) {
            if (i == 0) {
                path.moveTo(xy[i] - rec.left, xy[i + 1] - rec.top);
            } else {
                path.lineTo(xy[i] - rec.left, xy[i + 1] - rec.top);
            }
        }
        path.close();
        c.drawPath(path, paint);
        return bitmap;
    }

    private static Rect calcSearchRange(Rect mRoiBounds, int imgWidth, int imgHeight) {
        int xmin = Math.max(0, (int) (mRoiBounds.left - mRoiBounds.width() * 2));
        int ymin = Math.max(0, (int) (mRoiBounds.top - mRoiBounds.height() * 2));
        int xmax = (int) (mRoiBounds.right + mRoiBounds.width() * 2);
        int ymax = (int) (mRoiBounds.bottom + mRoiBounds.height() * 2);
        xmax = Math.min(imgWidth, xmax);
        ymax = Math.min(imgHeight, ymax);
        xmax = Math.max(0, xmax);
        ymax = Math.max(0, ymax);
        return new Rect(xmin, ymin, xmax, ymax);
    }
}

File: frameworks/rs/tests/java_api/HealingBrush/src/rs/example/android/com/healingbrush/find_region.rs
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#pragma version(1)
#pragma rs java_package_name(com.example.android.rs.sample)
#pragma rs_fp_relaxed

rs_allocation image;
rs_allocation border_values; // float3
rs_allocation border_coords; //int2

int borderLength;
/**
  This describes the algorithm that uses the code
  This takes a float2 polygon and an image and searches a best fit polygon in the image

  int2  __attribute__((invoke_cpu)) findBestFit(rs_allocation poly, rs_allocation image) {
    int search_range_width;
    int search_range_height;
    borderLength = rsAllocationGetDimX(poly);
    int width = rsAllocationGetDimX(image);
    int height = rsAllocationGetDimY(image);

    border_coords = alloc_int2(borderLength, 0, 0);
    foreach(toInt, poly , border_coords);
    border_values = alloc_float3(borderLength, 0, 0);
    forEach_extractBorder(border_coords , border_values);

    fit = alloc_float(search_range_width, search_range_width, 0);
    borderLength = size;

    forEach_bordercorrelation(image);
    int2 fit_pos = gfindMin(fit, cx, cy, cwidth, cheight);
    return fit_pos;
 }

void findRegion(rs_allocation coord, rs_allocation image) {
  borderLength = rsAllocationGetDimX(coord);
  int width = rsAllocationGetDimX(image);
  int height = rsAllocationGetDimY(image);
}

static int4 padRegionRect(float4 rec) {
  int width = rec.z;
  int height = rec.w;
  int  mWidth = (((int) (8 + width)) & ~3); // bounding rectangle that is a power of 8 big
  int  mHeight = (((int) (8 + height)) & ~3);
  int4 out = {(int)(rec.x - 1 ),(int)(rec.y - 1), width, height};
  return out;
}


*/


static float4 gcalcBounds(rs_allocation xy) {
  int len = rsAllocationGetDimX(xy);
  float2 min_xy = rsGetElementAt_float(xy, 0);
  float2 max_xy = min_xy;
  for (int i = 0; i < len; i += 2) {
    float2 v = rsGetElementAt_float2(xy, i);
    min_xy = min(min_xy, v);
    max_xy = max(max_xy, v);
  }
  max_xy-=min_xy;
  float4 out = {min_xy.x, min_xy.y, max_xy.x, max_xy.y};
  return out;
}

void calcBounds(rs_allocation xy, rs_allocation rect) {
  float4 r = gcalcBounds(xy);
  rsSetElementAt_float2(rect, r.xy, 0);
  rsSetElementAt_float2(rect, r.zw, 1);
}


float3 __attribute__((kernel))extractBorder(int2 in) {
   return convert_float3(rsGetElementAt_uchar4(image, in.x, in.y).xyz);
}

int2 __attribute__((kernel)) toInt(float2 in) {
  int2 out = {(int)in.x, (int) in.y};
  return out;
}


static int2 gfindMin(rs_allocation fit, int cx, int cy, int cwidth, int cheight) {
  int w = rsAllocationGetDimX(fit);
  int h = rsAllocationGetDimY(fit);
  float minFit = rsGetElementAt_float(fit, 0, 0);
  int2 fit_pos = {0, 0};
  int reg_minx = cx;
  int reg_miny = cy;
  int mWidth =  cwidth;
  int mHeight =  cheight;
  int reg_width = mWidth;
  int reg_height = mHeight;
  int reg_maxx = reg_minx + mWidth;
  int reg_maxy = reg_miny + mHeight;
  for (int y = 0; y < h - mHeight; y++) {

    for (int x = 0; x < w - mWidth; x++) {
      if (!(x > reg_maxx || x + reg_width < reg_minx || y > reg_maxy
          || y + reg_height < reg_miny)) {
        continue;
      }
      float v = rsGetElementAt_float(fit, x, y);
      if (v < minFit) {
        minFit = v;
        fit_pos.x = x;
        fit_pos.y = y;
      }
    }
  }

  return fit_pos;
}

void findMin(rs_allocation fit, rs_allocation fit_max,int cx, int cy, int cwidth, int cheight) {
  int2 fit_pos = gfindMin(fit, cx, cy, cwidth, cheight);
  rsSetElementAt_int2(fit_max, fit_pos, 0);
}

int imagePosX;
int imagePosY;

float __attribute__((kernel)) bordercorrelation(uint32_t x, uint32_t y) {
  float sum = 0;
  int dx = x-imagePosX;
  int dy = y-imagePosY;
  for(int i = 0 ; i < borderLength; i++) {
    int2  coord = rsGetElementAt_int2(border_coords, i);
    float3 orig = convert_float3(rsGetElementAt_uchar4(image, coord.x + dx, coord.y + dy).xyz);
    float3 candidate = rsGetElementAt_float3(border_values, i).xyz;
    sum += distance(orig, candidate);
  }
  return sum;
}

